src/Api.js

```diff
@@ -89,13 +95,97 @@ class Api {
 	}
 
 	/**
-	 * Get the WhoColor data for a given wiki page.
+	 * Extract token and user IDs from a WikiWho span element with `id='token-X'` and
+	 * `class='token-editor-Y'` attributes.
 	 *
-	 * @param  {string} wikiUrl URL of the wiki page that we want to analyze.
+	 * @param {HTMLElement} element
+	 * @return {int[]|boolean[]} A two-element array: 0 = token_id, 1 = user_id.
+	 */
+	getIdsFromElement( element ) {
+		const out = [ false, false ],
+			tokenMatches = element.getAttribute( 'id' ).match( /token-(\d+)/ ),
+			editorMatches = element.getAttribute( 'class' ).match( /token-editor-(\d+)/ );
+		if ( tokenMatches && tokenMatches[ 1 ] ) {
+			out[ 0 ] = parseInt( tokenMatches[ 1 ] );
+		}
+		if ( editorMatches && editorMatches[ 1 ] ) {
+			out[ 1 ] = parseInt( editorMatches[ 1 ] );
+		}
+		return out;
+	}
+
+	/**
+	 * Get the WikiWho replacement for `.mw-parser-output` HTML.
+	 * @return {string}
+	 */
+	getReplacementHtml() {
+		return this.results.extended_html;
+	}
+
+	/**
+	 * Get user and revision information for a given token.
+	 *
+	 * @param {int} tokenId
+	 * @return {{revisionId: *, score: *, userId: *, username: *, revisionTime: *}|boolean}
+	 */
+	getTokenInfo( tokenId ) {
+		let revId, revision, username, score;
+
+		// Get the token information. results.tokens structure:
+		// [ [ conflict_score, str, o_rev_id, in, out, editor/class_name, age ], ... ]
+		// e.g. Array(7) [ 0, "indicate", 769691068, [], [], "18201938", 76652371.587203 ]
+		const token = this.results.tokens[ tokenId ] || false;
+		if ( !token ) {
+			return false;
+		}
+
+		// Get revision information. results.revisions structure:
+		// { rev_id: [ timestamp, parent_rev_id, user_id, editor_name ], ... }
+		// e.g. Array(4) [ "2017-03-11T02:12:47Z", 769315355, "18201938", "Biogeographist" ]
+		revId = token[ 2 ];
+		revision = this.results.revisions[ revId ];
+		username = revision[ 3 ];
+
+		// Get the user's edit score (percentage of content edited).
+		// results.present_editors structure:
+		// [ [ username, user_id, score ], ... ]
+		for ( let i = 0; i < this.results.present_editors.length; i++ ) {
+			if ( this.results.present_editors[ i ][ 0 ] === username ) {
+				score = parseFloat( this.results.present_editors[ i ][ 2 ] ).toFixed( 1 );
+				break;
+			}
+		}
+
+		// Put it all together.
+		return {
+			username: username,
+			userId: token[ 5 ],
+			revisionId: revId,
+			revisionTime: new Date( revision[ 0 ] ),
+			score: score
+		};
+	}
+
+	/**
+	 * Get the WikiWho data for a given wiki page.
+	 *
+	 * @param {string} wikiUrl URL of the wiki page that we want to analyze.
 	 * @return {Promise}
 	 */
 	getData( wikiUrl ) {
-		return $.getJSON( this.getAjaxURL( wikiUrl ) );
+		const api = this,
+			request = $.getJSON( this.getAjaxURL( wikiUrl ) );
+		request.done( function ( result ) {
```

Comment:
To clarify: Attaching a .then will work in the sense that the code inside that `then` or `fail` would *run*, but it has no promise of order like a chained `then` does. If we continue with using this promise as a sort of flag ("we got the data, doesn't matter if the rest of the stuff works later/earlier") it might matter less, but if we ever need to order the responses, it may matter more.

Reply:
Makes sense. I've fixed it up.